### DISCUSSION TITLE
Access document file path within Lua filter

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#6916] [Q&A] [answered]
### Description

I have a document, `example.qmd` inside a Quarto project that is running a lua filter called `my_fil.lua`. Within the filter, is there any way to access the path to that file relative to the project file?

In the following example, I'd like to have access to `notes/example.qmd`.

```
- _extensions
  - my-fil
    - _extension.yml
    - my-fil.lua
- _quarto.yml
- notes
  - example.qmd
```

**Comment by jimjam-slam** [comment]
@andrewpbray For pre- and post-render scripts Quarto makes [ceertain environment variables](https://quarto.org/docs/projects/scripts.html#pre-and-post-render) available, but for a filter, maybe [`PANDOC_STATE.input_files`](https://pandoc.org/lua-filters.html#global-variables) will do the trick?

**Reply by mcanouil** [reply]
Indeed, there are a bunch of environment variables you can have access to (see <https://github.com/mcanouil/quarto-lua-env>).
Note that the environment variable mentioned is not Quarto input, but Pandoc input, you might prefer `quarto.doc.input_file`.

**Comment by cderv** [comment, marked as answer]
A few items for this purpose have been added in Quarto Lua api in version 1.2
https://github.com/quarto-dev/quarto-cli/blob/b60509d5b4017dd11dc791310af1a9cb152285af/news/changelog-1.2.md?plain=1#L33-L38
`quarto.doc.input_file` mentioned above ( https://github.com/quarto-dev/quarto-cli/discussions/6916#discussioncomment-7054384) is one of them

We still need to document all of them - so sorry to make it hard to find
* https://github.com/quarto-dev/quarto-cli/issues/4499

For now, reading the source is the best doc 
https://github.com/quarto-dev/quarto-cli/blob/b60509d5b4017dd11dc791310af1a9cb152285af/src/resources/pandoc/datadir/init.lua#L1916-L2084

**Reply by andrewpbray** [reply]
Hey, terrific: `quarto.doc.input_file` and `quarto.project.directory` get me exactly what I need.

Thanks!

**Reply by andrewpbray** [reply]
While working on a related project, I bumped into an issue that brought me back to this thread (and the source that you linked).

I'm trying to rename the output file (`-o`) within a lua filter. Based on looking at the source, I thought I'd try this:
```
PANDOC_STATE['output_file'] =  "foo.json"
```
Alas, it didn't work.
```
Error running filter /Applications/quarto/share/filters/main.lua:
'output_file' is a read-only property.
```
It may be that this simply isn't possible in this manner. If so I can brainstorm other ways to skin this cat.

**Reply by jimjam-slam** [reply]
Post-render project script to rename the output file? 🤔

**Reply by andrewpbray** [reply]
Indeed this is wrapped in a project, so that may work.... intriguing. Might work! I'd like this renaming to be subject to document-level metadata (effectively: `add-suffix: true`), and it'd like it to not overwrite the output that is under the original file name. I could conceivably do:

1. in pre-render script reach through docs and look for relevant metadata.
2. if `add-suffix: true` rename the file.
3. do all the renders.
4. in post-render script undo steps 2 and 1.

Thanks for the suggestion!

**Reply by cderv** [reply]
> It may be that this simply isn't possible in this manner. If so I can brainstorm other ways to skin this cat.

Yes I don't think Pandoc will allow that. They cleary says `PANDOC_STATE` is read only (https://pandoc.org/lua-filters.html#global-variables) 
> The state shared by all readers and writers. It is used by pandoc to collect and pass information. The value of this variable is of type [CommonState](https://pandoc.org/lua-filters.html#type-commonstate) and is read-only.

To be able to do that at the document level, it would need to be a Quarto feature. A feature like allowing to tweak some post process Quarto behavior from within Lua filter (like renaming a file by setting `output-file` YAML config). 

Personally I think that all path handling like this are best done with external processing like you are thinking off.
